core/src/main/java/org/apache/calcite/sql/validate/implicit/AbstractTypeCoercion.java

```diff
@@ -258,7 +258,7 @@ protected boolean needToCast(SqlValidatorScope scope, SqlNode node,
     RelDataType fromType = validator.deriveType(scope, node);
     // This depends on the fact that type validate happens before coercion.
     // We do not have inferred type for some node, i.e. LOCALTIME.
```

Comment:
It seems that judging the type of node, null and SqlTypeName.NUL are not equivalent in my opinion. Do we need to improve this part of the comment?

Reply:
This repair method is not comprehensive enough. I will check again.